fix(compiler): throw error for duplicate template references

crisbeto opened the pull request:
Adds an error if a reference is used more than once on the same element (e.g. `<div #a #a>`). We used to have this error in ViewEngine, but it wasn't ported over to Ivy.

Fixes #40536.

Airblader commented:
@crisbeto What about the other case mentioned in the ticket where the reference is not on the same element, but on two?

If I understand correctly this also fixes that one, but no test for it has been added.

crisbeto commented:
It looks like I missed the other case and this code only checked within the same element. I've updated the code to check across elements as well.

JoostK commented on packages/compiler/src/render3/r3_template_transform.ts:
```diff
@@ -75,6 +75,7 @@ class HtmlAstToIvyAst implements html.Visitor {
   styles: string[] = [];
   styleUrls: string[] = [];
   ngContentSelectors: string[] = [];
+  private allReferenceNames = new Set<string>();
```
Reference names are scoped to their containing template; whereas it looks like this tracks a global set. The VE template parser appears to be buggy here as well, in that it only checks for duplicated variable declarations in the root, not in nested templates. E.g. this correctly succeeds in VE:

```html
<div *ngIf="true" #divRef></div>
<div #divRef></div>
```

but this should be rejected, but is incorrectly accepted:

```html
<div *ngIf="true" #divRef #divRef></div>
```

Also note that the similar operation of reporting duplicate template variables (which is different to reference declarations, to be clear) is currently checked during type-checking:

https://github.com/angular/angular/blob/d466db8285920347972cec53250e0937f07726c7/packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts#L1142-L1152

That's just for reference, though; I think it's fine to report this error from within the template transform.

crisbeto replied:
My understanding was that we create one `HtmlToIvyAst` per parsed file so the registry shouldn't be global, or am I mistaken? For reference, here's where we create it: https://github.com/angular/angular/blob/master/packages/compiler/src/render3/r3_template_transform.ts#L58

JoostK replied:
My usage of global was a bit poor; I meant to say "for a single component template". The single set is a problem for e.g:

```html
<div *ngIf="true" #divRef></div>
<div #divRef></div>
```

as the first `divRef` occurs within an `ng-template` which creates a new scope, so the two declarations are actually not conflicting.

crisbeto replied:
I see what you mean now. I suppose that we can make it work, but the question is whether it's worth the trouble. In my opinion, even though the names won't conflict at runtime, it's cleaner if they're all unique anyway.

crisbeto commented:
Now that I'm looking at it again, it seems like Ivy explicitly supports multiple local references within the same template. See https://github.com/angular/angular/blob/master/packages/core/test/acceptance/query_spec.ts#L64. I'll roll back the changes so they only verify that the references are unique within a particular element.

crisbeto committed:
fix(compiler): throw error for duplicate template references

Adds an error if a reference is used more than once on the same element (e.g. `<div #a #a>`).
We used to have this error in ViewEngine, but it wasn't ported over to Ivy.

Fixes #40536.

alxhub commented on packages/compiler/src/render3/r3_template_transform.ts:
```diff
@@ -458,6 +458,8 @@ class HtmlAstToIvyAst implements html.Visitor {
     } else if (identifier.length === 0) {
       this.reportError(`Reference does not have a name`, sourceSpan);
+    } else if (references.some(reference => reference.name === identifier)) {
+      this.reportError(`Reference "#${identifier}" is defined several times`, sourceSpan);
```
```suggestion
      this.reportError(`Reference "#${identifier}" is defined more than once`, sourceSpan);
```

crisbeto committed:
fixup! fix(compiler): throw error for duplicate template references

AndrewKushnir commented:
[Presubmit](http://test/OCL:356378090:BASE:356378105:1612830035269:69e31935).
